wagtail/wagtailcore/models.py

```diff
@@ -1277,17 +1283,184 @@ def serve_password_required_response(self, request, form, action_url):
+    Converts the output of get_explorable_page_paths() to a Q object which will filter a QuerySet down to the
+    appropriate set of Pages.
+    """
+    path_Qs = Q()
+    for path in permitted_paths:
+        path_Qs = path_Qs | Q(path__startswith=path)
+    for ancestor in required_ancestors:
+        path_Qs = path_Qs | Q(path=ancestor)
+    return path_Qs
+
+
+def clear_explorable_paths_cache(user=None):
+    """
+    Whenever a Page's path changes, or a Group's permissions change, or a User's Groups change, some or all of the
+    cached sets of explorable pages may now be invalid. Unfortunately, Django's caching system has very limited
+    ability to selectively clear cached data. So to avoid unnecessarily clearing unrelated cache data, we have to
+    be a bit brute forcey.
+
+    Pass in a user object to clear just that user's cached permitted path data. Otherwise, all cached permitted path
+    data will be cleared.
+    """
+    if user:
+        cache.delete(EXPLORABLE_PATHS_PREFIX.format(user.pk))
+    else:
+        # Delete any cached explorable paths that might exist for every user in the system.
+        cache.delete_many(
+            EXPLORABLE_PATHS_PREFIX.format(uid) for uid in get_user_model().objects.values_list('pk', flat=True)
+        )
+
+
+def get_navigation_menu_items(user):
     # Get all pages that appear in the navigation menu: ones which have children,
     # or are at the top-level (this rule required so that an empty site out-of-the-box has a working menu)
     pages = Page.objects.filter(Q(depth=2) | Q(numchild__gt=0)).order_by('path')
 
-    # Turn this into a tree structure:
+    # We need to treat the top of the user's explorable tree as if it were depth=2, or the depth_list algorithm won't
+    # work. So if a user isn't permitted to see pages at or below depth 2, we must change the apparent depth of each
+    # of their explorable pages.
+    depth_change = 0
+    if not user.is_superuser:
+        # For backward compatibility of the explorer_nav template tag, we accept "None" for the user argument, in which
+        # case we can't do permissions-based restriction.
+        # TODO: Is this check even necessary? I'm not sure what the backward compatibility rules are, or if the
+        # {% explorer_nav %} tag was ever used in the first place.
+        if user:
+            permitted_paths, required_ancestors = get_explorable_page_paths(user)
+            if permitted_paths:
+                # Limit the listing to only the user's explorable pages. Always include the Root page, even if it's
+                # not an explorable path, since it's required for the depth_list algorithm.
+                pages = pages.filter(
+                    convert_explorable_paths_to_Q(permitted_paths, required_ancestors) | Q(depth=1)
+                )
+
+                # Find the length of the shortest explorable path.
+                shortest_path_len = len(min(permitted_paths + required_ancestors, key=len))
```

Comment:
Oh hey, you're right! I wrote the original version of that algorithm before I decided to forcibly put the CCA at index 0 in required_ancestors. I'll fix that.